parquet/src/main/java/org/apache/iceberg/parquet/ParquetMetrics.java

```diff
@@ -0,0 +1,639 @@
+        long valueCount = 0;
+        long nullCount = 0;
+
+        for (ColumnChunkMetaData column : columns.get(path)) {
+          Statistics<?> stats = column.getStatistics();
+          if (stats == null || stats.isEmpty()) {
+            // the null count is unknown
+            return null;
+          }
+
+          boolean hasOnlyNullVariants;
+          if (stats.hasNonNullValue()) {
+            hasOnlyNullVariants =
+                Variants.isNull(ByteBuffer.wrap(stats.getMinBytes()))
+                    && Variants.isNull(ByteBuffer.wrap(stats.getMaxBytes()));
+          } else {
+            // use the null count because the min/max were not defined
+            hasOnlyNullVariants = false;
+          }
+
+          valueCount += column.getValueCount();
+          nullCount += hasOnlyNullVariants ? column.getValueCount() : stats.getNumNulls();
+        }
+
+        return new ParquetVariantUtil.VariantMetrics(valueCount, nullCount);
+      }
+
+      @SuppressWarnings("CyclomaticComplexity")
+      private <T> ParquetVariantUtil.VariantMetrics metrics(PrimitiveType primitive) {
+        PhysicalType variantType = ParquetVariantUtil.convert(primitive);
+        if (null == variantType) {
+          // the type could not be converted and is either invalid or unsupported
+          return null;
+        }
+
+        ColumnPath path = ColumnPath.get(currentPath().toArray(String[]::new));
+        Comparator<T> comparator = ParquetVariantUtil.comparator(variantType);
+        int scale = ParquetVariantUtil.scale(primitive);
+        long valueCount = 0;
+        long nullCount = 0;
+        T lowerBound = null;
+        T upperBound = null;
+
+        for (ColumnChunkMetaData column : columns.get(path)) {
+          Statistics<?> stats = column.getStatistics();
+          if (stats == null || stats.isEmpty()) {
+            return null;
+          }
+
+          nullCount += stats.getNumNulls();
+          valueCount += column.getValueCount();
+
+          if (stats.hasNonNullValue()) {
+            T chunkMin = ParquetVariantUtil.convertValue(variantType, scale, stats.genericGetMin());
+            if (lowerBound == null || comparator.compare(chunkMin, lowerBound) < 0) {
+              lowerBound = chunkMin;
+            }
+
+            T chunkMax = ParquetVariantUtil.convertValue(variantType, scale, stats.genericGetMax());
+            if (upperBound == null || comparator.compare(chunkMax, upperBound) > 0) {
+              upperBound = chunkMax;
+            }
+          }
+        }
+
+        if (NaNUtil.isNaN(lowerBound) || NaNUtil.isNaN(upperBound)) {
+          return null;
+        }
+
+        if (lowerBound != null && upperBound != null) {
+          VariantValue lower = Variants.of(variantType, lowerBound);
+          VariantValue upper = Variants.of(variantType, upperBound);
+          return new ParquetVariantUtil.VariantMetrics(valueCount, nullCount, lower, upper);
```

Comment:
It's technically possible that there's a lower bound but no upper bound (or vice versa) right? Should we just always be using the 4-arg constructor and passing null for lower/upper respectively for those cases.

Reply:
No nvm, in this path we'd be guaranteed that both lower/upper bound would be populated or neither would be populated.